src/ADAL.PCL.Desktop/DeviceID.cs

```diff
@@ -0,0 +1,147 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System;
+using System.Drawing;
+using System.Collections;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Windows.Forms;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Microsoft.IdentityModel.Clients.ActiveDirectory
+{
+    /// <summary>
+    /// Summary description for DeviceID.
+    /// </summary>
+    public class DeviceID
+    {
+
+        internal static string strDeviceID { get; set; }
+
+        public DeviceID()
+        {
+            strDeviceID = GetDeviceID();
+        }
+
+
+        private static Int32 METHOD_BUFFERED = 0;
+        private static Int32 FILE_ANY_ACCESS = 0;
+        private static Int32 FILE_DEVICE_HAL = 0x00000101;
+
+        private const Int32 ERROR_NOT_SUPPORTED = 0x32;
+        private const Int32 ERROR_INSUFFICIENT_BUFFER = 0x7A;
+
+        private static Int32 IOCTL_HAL_GET_DEVICEID =
+            ((FILE_DEVICE_HAL) << 16) | ((FILE_ANY_ACCESS) << 14)
+            | ((21) << 2) | (METHOD_BUFFERED);
+
+        [DllImport("coredll.dll", SetLastError = true)]
+        private static extern bool KernelIoControl(Int32 dwIoControlCode,
+            IntPtr lpInBuf, Int32 nInBufSize, byte[] lpOutBuf,
+            Int32 nOutBufSize, ref Int32 lpBytesReturned);
+
+        internal static string GetDeviceID()
+        {
+            // Initialize the output buffer to the size of a 
+            // Win32 DEVICE_ID structure.
+            byte[] outbuff = new byte[20];
+            Int32 dwOutBytes;
+            bool done = false;
+
+            Int32 nBuffSize = outbuff.Length;
+
+            // Set DEVICEID.dwSize to size of buffer.  Some platforms look at
+            // this field rather than the nOutBufSize param of KernelIoControl
+            // when determining if the buffer is large enough.
+            BitConverter.GetBytes(nBuffSize).CopyTo(outbuff, 0);
+            dwOutBytes = 0;
+
+            // Loop until the device ID is retrieved or an error occurs.
+            while (!done)
```

Comment:
According to the C# documentation [https://msdn.microsoft.com/en-us/library/51y09td4(v=vs.71).aspx](https://msdn.microsoft.com/en-us/library/51y09td4(v=vs.71).aspx) , when the new operator fails it will throw OutOfMemoryException.I have added a try/catch for this. It would exit the loop on the occurrence of OutOfMemory exception with deviceID = null.

---
In reply to: [73685336](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/494#discussion_r73685336) [](ancestors = 73685336)